cvat/apps/engine/migrations/0076_remove_storage_cloud_storage_id_and_more.py

```diff
@@ -0,0 +1,44 @@
+# Generated by Django 4.2.6 on 2023-11-17 10:10
+
+from django.db import migrations, models
+import django.db.models.deletion
+from cvat.apps.engine.models import Location
+
+def manually_remove_outdated_relations(apps, schema_editor):
+    Storage = apps.get_model("engine", "Storage")
+    CloudStorage = apps.get_model("engine", "CloudStorage")
+    db_storages = Storage.objects.filter(cloud_storage_id__isnull=False)
+    db_storages_to_update = []
+
+    for db_storage in db_storages:
+        if not CloudStorage.objects.filter(pk=db_storage.cloud_storage_id).exists():
+            db_storage.cloud_storage_id = None
+            db_storage.location = Location.LOCAL
+            db_storages_to_update.append(db_storage)
+
+    if db_storages_to_update:
+        Storage.objects.bulk_update(db_storages_to_update, fields=['cloud_storage_id', 'location'], batch_size=500)
```

Comment:
A couple questions about this:

1. I notice that here you're updating the invalid `Storage` objects by resetting the `location` and `cloud_storage_id` fields. But that's not what happens with any future `Storage` objects whose corresponding `CloudStorage` is deleted - since you used `on_delete=CASCADE`, such `Storage` objects will be deleted entirely. To maintain consistency, wouldn't it make sense to delete the `Storage` objects here, as well?

2. I think it should be possible to do this deletion entirely within the database. Something like this should work: `Storage.objects.filter(Exists(CloudStorage.objects.filter(pk=OuterRef('cloud_storage_id')))).delete()`. This should be simpler and more efficient. You could also replace `delete` with `update` if you disagree with item 1.

Reply:
1. Yes, you are right
2. Thanks